File: baekjoon/17135/17135.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public class Main {
    private static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    private static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    private static int N, M, D;
    private static int[][] matrix, tempMatrix;
    private static int maxAttackedEnemyCount;

    public static void main(String[] args) throws IOException {
        int[] input = toIntArray(reader.readLine().split(" "));

        N = input[0];
        M = input[1];
        D = input[2];

        matrix = new int[N + 1][M];

        for (int i = 0; i < N; i++) {
            matrix[i] = toIntArray(reader.readLine().split(" "));
        }

        solve();

        writer.write(String.format("%d\n", maxAttackedEnemyCount));
        writer.flush();

        writer.close();
    }

    private static int[] toIntArray(String[] strings) {
        int[] ints = new int[strings.length];

        for (int i = 0; i < strings.length; i++) {
            ints[i] = Integer.parseInt(strings[i]);
        }

        return ints;
    }

    private static void solve() {
        for (int i = 0; i < M - 2; i++) {
            for (int j = i + 1; j < M - 1; j++) {
                for (int k = j + 1; k < M; k++) {
                    tempMatrix = new int[N][M];
                    copyMatrix(matrix, tempMatrix);

                    int firstArcherCol = i;
                    int secondArcherCol = j;
                    int thirdArcherCol = k;

                    int attackedEnemyCount = 0;

                    while (isEnemyExist()) {
                        Set<Location> attackedLocations = new HashSet<>();
                        attack(firstArcherCol, attackedLocations);
                        attack(secondArcherCol, attackedLocations);
                        attack(thirdArcherCol, attackedLocations);

                        for (Location attackedLocation : attackedLocations) {
                            tempMatrix[attackedLocation.row][attackedLocation.col] = 0;
                        }
                        attackedEnemyCount += attackedLocations.size();

                        moveEnemy();
                    }

                    if (attackedEnemyCount > maxAttackedEnemyCount) {
                        maxAttackedEnemyCount = attackedEnemyCount;
                    }
                }
            }
        }
    }

    private static void copyMatrix(int[][] matrix1, int[][] matrix2) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                matrix2[i][j] = matrix1[i][j];
            }
        }
    }

    private static boolean isEnemyExist() {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (tempMatrix[i][j] == 1) {
                    return true;
                }
            }
        }
        return false;
    }

    private static void attack(int archerCol, Set<Location> attackedLocations) {
        Location minDistanceEnemyLocation = null;
        int minDistance = Integer.MAX_VALUE;

        for (int j = 0; j < M; j++) {
            for (int i = 0; i < N; i++) {
                if (tempMatrix[i][j] == 1 && distance(N, archerCol, i, j) <= D) {
                    if (minDistanceEnemyLocation == null) {
                        minDistance = distance(N, archerCol, i, j);
                        minDistanceEnemyLocation = new Location(i, j);
                        continue;
                    }

                    int distance = distance(N, archerCol, i, j);
                    if (distance < minDistance) {
                        minDistance = distance;
                        minDistanceEnemyLocation = new Location(i, j);
                    }
                }
            }
        }

        if (minDistanceEnemyLocation != null) {
            attackedLocations.add(minDistanceEnemyLocation);
        }
    }

    private static int distance(int row1, int col1, int row2, int col2) {
        return Math.abs(row1 - row2) + Math.abs(col1 - col2);
    }

    private static void moveEnemy() {
        for (int i = N - 1; i > 0; i--) {
            tempMatrix[i] = tempMatrix[i - 1];
        }
        tempMatrix[0] = new int[M];
    }

    private static class Location {
        int row, col;

        public Location(int row, int col) {
            this.row = row;
            this.col = col;
        }

        public boolean equals(Object object) {
            if (object == null) {
                return false;
            }
            if (object == this) {
                return true;
            }
            if (!(object instanceof Location)) {
                return false;
            }

            return row == ((Location) object).row && col == ((Location) object).col;
        }

        public int hashCode() {
            return Objects.hash(row, col);
        }
    }
}

File: baekjoon/17135/17135.py
import sys

matrix = None
temp_matrix = None
N = None
M = None
D = None
max_attacked_enemy_count = 0

class Location:
    def __init__(self, row, col):
        self.row = row
        self.col = col

    def __eq__(self, other):
        if other is None:
            return False
        if other is self:
            return True
        if type(other) is not Location:
            return False

        return self.row == other.row and self.col == other.col

    def __hash__(self):
        return hash((self.row, self.col))

def solve():
    global matrix, temp_matrix, N, M, D, max_attacked_enemy_count

    for i in range(0, M - 2):
        for j in range(i + 1, M - 1):
            for k in range(j + 1, M):
                temp_matrix = [[0] * M for _ in range(N)]
                copy_matrix(matrix, temp_matrix)

                first_archer_col = i
                second_archer_col = j
                third_archer_col = k

                attacked_enemy_count = 0

                while is_enemy_exist():
                    attacked_locations = set()
                    attack(first_archer_col, attacked_locations)
                    attack(second_archer_col, attacked_locations)
                    attack(third_archer_col, attacked_locations)

                    for attacked_location in attacked_locations:
                        temp_matrix[attacked_location.row][attacked_location.col] = 0

                    attacked_enemy_count += len(attacked_locations)

                    move_enemy()

                if attacked_enemy_count > max_attacked_enemy_count:
                    max_attacked_enemy_count = attacked_enemy_count


def copy_matrix(matrix1, matrix2):
    global matrix, temp_matrix, N, M, D, max_attacked_enemy_count

    for i in range(N):
        for j in range(M):
            matrix2[i][j] = matrix1[i][j]

def is_enemy_exist():
    global matrix, temp_matrix, N, M, D, max_attacked_enemy_count

    for i in range(N):
        for j in range(M):
            if temp_matrix[i][j] == 1:
                return True

    return False

def attack(archer_col, attacked_locations):
    global matrix, temp_matrix, N, M, D, max_attacked_enemy_count

    min_distance_enemy_location = None
    min_distance = None

    for j in range(M):
        for i in range(N):
            if temp_matrix[i][j] == 1 and get_distance(N, archer_col, i, j) <= D:
                if min_distance_enemy_location is None:
                    min_distance = get_distance(N, archer_col, i, j)
                    min_distance_enemy_location = Location(i, j)
                    continue

                distance = get_distance(N, archer_col, i, j)
                if distance < min_distance:
                    min_distance = distance
                    min_distance_enemy_location = Location(i, j)

    if min_distance_enemy_location is not None:
        attacked_locations.add(min_distance_enemy_location)


def get_distance(row1, col1, row2, col2):
    global matrix, temp_matrix, N, M, D, max_attacked_enemy_count

    return abs(row1 - row2) + abs(col1 - col2)

def move_enemy():
    global matrix, temp_matrix, N, M, D, max_attacked_enemy_count

    for i in range(N - 1, 0, -1):
        temp_matrix[i] = temp_matrix[i - 1]

    temp_matrix[0] = [0] * M

if __name__ == '__main__':
    N, M, D = map(int, sys.stdin.readline().split())
    matrix = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]

    solve()

    print(max_attacked_enemy_count)
